cmd/query/app/static_handler.go

```diff
@@ -106,14 +107,10 @@ func loadAndEnrichIndexHTML(open func(string) (http.File, error), options Static
 	}
 	// replace UI config
 	configString := "JAEGER_CONFIG = DEFAULT_CONFIG"
-	if config, err := loadUIConfig(options.UIConfigPath); err != nil {
+	if configBytes, err := loadUIConfig(options.UIConfigPath); err != nil {
 		return nil, err
-	} else if config != nil {
-		// TODO if we want to support other config formats like YAML, we need to normalize `config` to be
-		// suitable for json.Marshal(). For example, YAML parser may return a map that has keys of type
-		// interface{}, and json.Marshal() is unable to serialize it.
-		bytes, _ := json.Marshal(config)
-		configString = fmt.Sprintf("JAEGER_CONFIG = %v", string(bytes))
+	} else if configBytes != nil {
+		configString = fmt.Sprintf("JAEGER_CONFIG = %v", string(configBytes))
```

Comment:
This doesn't look right. In case of JS file that contains a function there needs to be a part of the template that invokes that function and assigns its _result_ to JAEGER_CONFIG ( or something similar).

Reply:
I would suggest taking a look at this line from the perspective of invoking the JS config file itself. 

My idea is next:
- we add a more strict JS file format validation like `/^function\s[a-zA-Z$_][a-zA-Z0-9$_]+(\s)?\(\)(\s)?{(.*\n)+}$/gi` - in this case we can rely on JS file and assume, that its plain JS function like
```
function LoadCodnfig(){
    console.log('Hello World!');
    alert('Im in Jaeger UI!');

    console.log('asdasd');

    return {
        callMe: () => {
            return Promise.resolve(setTimeout(() => {
                console.log('CALL ME!!!', 1000);
            }))
        },
        asd: 'sdf'
    }
}
```
- next, we add `()` at the moment of injecting config inside `index.html` and we have IIFE inside `index.html`
- in `jaeger-ui` we will add a cache for `getConfig()` like
```
diff --git a/packages/jaeger-ui/src/utils/config/get-config.tsx b/packages/jaeger-ui/src/utils/config/get-config.tsx
index 458ba0c..68dfd47 100644
--- a/packages/jaeger-ui/src/utils/config/get-config.tsx
+++ b/packages/jaeger-ui/src/utils/config/get-config.tsx
@@ -16,15 +16,21 @@ import _get from 'lodash/get';
 
 import processDeprecation from './process-deprecation';
 import defaultConfig, { deprecations } from '../../constants/default-config';
+import { Config } from '../../types/config';
 
 let haveWarnedFactoryFn = false;
 let haveWarnedDeprecations = false;
+let cachedConfig: null | Config = null;
 
 /**
  * Merge the embedded config from the query service (if present) with the
  * default config from `../../constants/default-config`.
  */
 export default function getConfig() {
+  if(cachedConfig) {
+    return cachedConfig;
+  }
+
   const getJaegerUiConfig = window.getJaegerUiConfig;
   if (typeof getJaegerUiConfig !== 'function') {
     if (!haveWarnedFactoryFn) {
@@ -52,6 +58,7 @@ export default function getConfig() {
       rv[key] = { ...defaultConfig[key], ...embedded[key] };
     }
   }
+  cachedConfig = rv;
   return rv;
 }
```

From a usage perspective, we can say, that JS config file is [Function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function) and we invoke it once the page loaded

Reply:
Could you show how index.html would be changed to handle JS file? Right now it looks like this, and the middle line is being replaced with the content of JSON file.
```
      function getJaegerUiConfig() {
        const DEFAULT_CONFIG = null;
        const JAEGER_CONFIG = DEFAULT_CONFIG;
        return JAEGER_CONFIG;
      }
```

Reply:
For JSON:
```
      function getJaegerUiConfig() {
        const DEFAULT_CONFIG = null;
        const JAEGER_CONFIG = {"a":1, "b":2};
        return JAEGER_CONFIG;
      }
```

For JS:
```
      function getJaegerUiConfig() {
        const DEFAULT_CONFIG = null;
        const JAEGER_CONFIG = function UIConfig() {
          return {"a":1, "b":2};
        }();
        return JAEGER_CONFIG;
      }
```

Reply:
I see, you're suggesting to piggy back on the string replacement that currently happens for `const JAEGER_CONFIG = DEFAULT_CONFIG;` line. While this is doable, I wonder if it puts a lot of limitations on what the actual config function can do. If all it does is return a static dictionary, then sure, but we may want to use it as a mechanism for loading plugins, etc. Wouldn't it be better if it was embedded at the top level, not inlined inside an expression?

What if we do this instead:
```javascript
// JAEGER_CONFIG_JS
// the line above may be replaced by user-provided JS file that should define a UIConfig function.

function getJaegerUiConfig() {
        if "UIConfig is not null and is a function" {
             return UIConfig();
        }
        const DEFAULT_CONFIG = null;
        const JAEGER_CONFIG = DEFAULT_CONFIG;
        return JAEGER_CONFIG;
}
```

Reply:
regarding this:

> let cachedConfig: null | Config = null;

I think there is a memoize module that we're using, could it be used instead of explicit caching?

Reply:
Do I understand you correctly?

JSON:
```
// JAEGER_CONFIG_JS
// the line above may be replaced by user-provided JS file that should define a UIConfig function.

function getJaegerUiConfig() {
        if(typeof window.UIConfig === 'function') {
             return UIConfig();
        }
        const DEFAULT_CONFIG = null;
        const JAEGER_CONFIG = {a:1, b:2};
        return JAEGER_CONFIG;
}
```

JS:
```
function UIConfig() {
   return {a:1, b:2}
}
// the line above may be replaced by user-provided JS file that should define a UIConfig function.

function getJaegerUiConfig() {
        if(typeof window.UIConfig === 'function') {
             return UIConfig();
        }
        const DEFAULT_CONFIG = null;
        const JAEGER_CONFIG = {a:1, b:2};
        return JAEGER_CONFIG;
}
```

Reply:
lgtm. any concerns?

Reply:
No concerns. Thanks for clarifying! 🤝  WIll update PR shortly

Reply:
bonus points if we can get rid of the commend during substitution (e.g. put it all on one line)